File: claude-code-agents-python/src/plugin_gen/scaffold.py
"""Generate a full plugin directory from a PluginSpec."""

from __future__ import annotations

import logging
from pathlib import Path

from ..models.plugin_spec import PluginSpec
from .agent_writer import write_agent
from .connectors_writer import write_connectors
from .hooks_writer import write_hooks
from .lsp_config import write_lsp_config
from .manifest import write_manifest
from .mcp_config import write_mcp_config
from .skill_writer import write_skill

logger = logging.getLogger(__name__)


def generate_plugin(spec: PluginSpec, output_dir: str | Path) -> Path:
    """Generate a complete plugin directory structure from a PluginSpec.

    Creates the following structure:
        <output_dir>/<plugin-name>/
            plugin.json
            skills/
                <skill-name>.md
                ...
            agents/
                <agent-name>.md
                ...
            connectors/
                mcp.json
            hooks/
                hooks.json
            .lsp/
                config.json

    Args:
        spec: The plugin specification to generate from.
        output_dir: Parent directory where the plugin directory will be created.

    Returns:
        Path to the generated plugin directory.
    """
    output_dir = Path(output_dir)
    plugin_dir = output_dir / spec.plugin_dir_name
    plugin_dir.mkdir(parents=True, exist_ok=True)

    logger.info("Generating plugin '%s' at %s", spec.name, plugin_dir)

    # Create subdirectories
    skills_dir = plugin_dir / "skills"
    skills_dir.mkdir(exist_ok=True)

    agents_dir = plugin_dir / "agents"
    agents_dir.mkdir(exist_ok=True)

    connectors_dir = plugin_dir / "connectors"
    connectors_dir.mkdir(exist_ok=True)

    hooks_dir = plugin_dir / "hooks"
    hooks_dir.mkdir(exist_ok=True)

    lsp_dir = plugin_dir / ".lsp"
    lsp_dir.mkdir(exist_ok=True)

    # Write skills
    for skill_spec in spec.skills:
        write_skill(skill_spec, skills_dir)
        logger.info("  Wrote skill: %s", skill_spec.full_name())

    # Write agents
    for agent_spec in spec.agents:
        write_agent(agent_spec, agents_dir)
        logger.info("  Wrote agent: %s", agent_spec.full_name())

    # Write connectors and MCP config
    if spec.connectors:
        write_connectors(spec.connectors, connectors_dir)
        write_mcp_config(spec.connectors, connectors_dir / "mcp.json")
        logger.info("  Wrote %d connectors", len(spec.connectors))

    # Write hooks
    if spec.hooks:
        write_hooks(spec.hooks, hooks_dir / "hooks.json")
        logger.info("  Wrote hooks for %d events", len(spec.hooks))

    # Write LSP config
    if spec.lsp_servers:
        write_lsp_config(spec.lsp_servers, lsp_dir / "config.json")
        logger.info("  Wrote LSP config for %d servers", len(spec.lsp_servers))

    # Write manifest
    write_manifest(spec, plugin_dir / "plugin.json")
    logger.info("  Wrote plugin.json manifest")

    logger.info(
        "Plugin generation complete: %d skills, %d agents, %d connectors",
        spec.skill_count,
        spec.agent_count,
        spec.connector_count,
    )
    return plugin_dir

File: claude-code-agents-typescript/src/plugin_gen/scaffold.ts
// src/plugin_gen/scaffold.ts — Main plugin scaffold generator
import * as fs from 'node:fs';
import * as path from 'node:path';
import type { PluginSpec } from '../models/plugin-spec.js';
import { pluginDirName } from '../models/plugin-spec.js';
import { writeManifest } from './manifest.js';
import { writeSkill } from './skill-writer.js';
import { writeAgent } from './agent-writer.js';
import { writeConnectors } from './connectors-writer.js';
import { writeHooks } from './hooks-writer.js';
import { writeLspConfig } from './lsp-config.js';
import { writeMcpConfig } from './mcp-config.js';

export function generatePlugin(spec: PluginSpec, outputDir: string): string {
  const dirName = pluginDirName(spec);
  const pluginDir = path.join(outputDir, dirName);

  // Create directory structure
  const dirs = [
    pluginDir,
    path.join(pluginDir, 'skills'),
    path.join(pluginDir, 'agents'),
    path.join(pluginDir, 'connectors'),
    path.join(pluginDir, 'hooks'),
  ];
  for (const dir of dirs) {
    fs.mkdirSync(dir, { recursive: true });
  }

  // Write manifest
  writeManifest(spec, pluginDir);

  // Write skills
  for (const skill of spec.skills) {
    writeSkill(skill, path.join(pluginDir, 'skills'));
  }

  // Write agents
  for (const agent of spec.agents) {
    writeAgent(agent, path.join(pluginDir, 'agents'));
  }

  // Write connectors
  if (spec.connectors.length > 0) {
    writeConnectors(spec.connectors, path.join(pluginDir, 'connectors'));
  }

  // Write hooks
  if (Object.keys(spec.hooks).length > 0) {
    writeHooks(spec.hooks, path.join(pluginDir, 'hooks'));
  }

  // Write LSP config
  if (spec.lspServers.length > 0) {
    writeLspConfig(spec.lspServers, pluginDir);
  }

  // Write MCP config
  if (spec.connectors.length > 0) {
    writeMcpConfig(spec.connectors, pluginDir);
  }

  return pluginDir;
}
